File: backend/api/routes/admin/locations.py
from fastapi import APIRouter, Depends, status

from backend.api.dependencies import get_location_service
from backend.api.schemas.location import (
    LocationCreate,
    LocationResponse,
    LocationUpdate,
)
from backend.services.location_service import LocationService


router = APIRouter(
    prefix="/locations",
    tags=["Admin — Locations"],
)


@router.post(
    "",
    response_model=int,
    status_code=status.HTTP_201_CREATED,
)
async def create_location(
    data: LocationCreate,
    service: LocationService = Depends(
        get_location_service,
    ),
):
    return await service.create(
        title=data.title,
        address=data.address,
        latitude=data.latitude,
        longitude=data.longitude,
        position=data.position,
        is_hidden=data.is_hidden,
    )


@router.get(
    "/{location_id}",
    response_model=LocationResponse,
)
async def get_location(
    location_id: int,
    service: LocationService = Depends(
        get_location_service,
    ),
):
    return await service.get_by_id(
        location_id,
    )


@router.get(
    "",
    response_model=list[LocationResponse],
)
async def get_locations(
    is_hidden: bool | None = None,
    service: LocationService = Depends(
        get_location_service,
    ),
):
    return await service.get_all(
        is_hidden=is_hidden,
    )


@router.put(
    "/{location_id}",
)
async def update_location(
    location_id: int,
    data: LocationUpdate,
    service: LocationService = Depends(
        get_location_service,
    ),
):
    await service.update(
        location_id=location_id,
        title=data.title,
        address=data.address,
        latitude=data.latitude,
        longitude=data.longitude,
        position=data.position,
        is_hidden=data.is_hidden,
    )

    return {
        "message": "Локация обновлена."
    }


@router.delete(
    "/{location_id}",
)
async def delete_location(
    location_id: int,
    service: LocationService = Depends(
        get_location_service,
    ),
):
    await service.delete(
        location_id,
    )

    return {
        "message": "Локация удалена."
    }

File: frontend/src/api/locations.ts
import { apiRequest } from './client'
import type {
  LocationCreate,
  LocationUpdate,
  LocationResponse,
  PublicLocationResponse,
} from '../types/location'

export function getPublicLocations(): Promise<PublicLocationResponse[]> {
  return apiRequest<PublicLocationResponse[]>('/locations')
}

export function getPublicLocation(
  locationId: number,
): Promise<PublicLocationResponse> {
  return apiRequest<PublicLocationResponse>(
    `/locations/${locationId}`,
  )
}

export function getAdminLocations(
  isHidden?: boolean,
): Promise<LocationResponse[]> {
  const query = isHidden === undefined
    ? ''
    : `?is_hidden=${isHidden}`

  return apiRequest<LocationResponse[]>(
    `/admin/locations${query}`,
  )
}

export function getAdminLocation(
  locationId: number,
): Promise<LocationResponse> {
  return apiRequest<LocationResponse>(
    `/admin/locations/${locationId}`,
  )
}

export function createLocation(
  data: LocationCreate,
): Promise<number> {
  return apiRequest<number>('/admin/locations', {
    method: 'POST',
    body: JSON.stringify(data),
  })
}

export function updateLocation(
  locationId: number,
  data: LocationUpdate,
): Promise<{ message: string }> {
  return apiRequest<{ message: string }>(
    `/admin/locations/${locationId}`,
    {
      method: 'PUT',
      body: JSON.stringify(data),
    },
  )
}

export function deleteLocation(
  locationId: number,
): Promise<{ message: string }> {
  return apiRequest<{ message: string }>(
    `/admin/locations/${locationId}`,
    {
      method: 'DELETE',
    },
  )
}